zcash_primitives/src/memo.rs

```diff
@@ -0,0 +1,412 @@
+    }
+
+    /// Returns the raw byte array containing the memo bytes, including null padding.
+    pub fn as_array(&self) -> &[u8; 512] {
+        &self.0
+    }
+
+    /// Returns a slice of the raw bytes, excluding null padding.
+    pub fn as_slice(&self) -> &[u8] {
+        let first_null = self
+            .0
+            .iter()
+            .enumerate()
+            .rev()
+            .find(|(_, &b)| b != 0)
+            .map(|(i, _)| i + 1)
+            .unwrap_or_default();
+
+        &self.0[..first_null]
+    }
+}
+
+/// Type-safe wrapper around String to enforce memo length requirements.
+#[derive(Clone, PartialEq)]
+pub struct TextMemo(String);
+
+impl From<TextMemo> for String {
+    fn from(memo: TextMemo) -> String {
+        memo.0
+    }
+}
+
+impl Deref for TextMemo {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &str {
+        self.0.deref()
+    }
+}
+
+/// An unencrypted memo received alongside a shielded note in a Zcash transaction.
+#[derive(Clone)]
+pub enum Memo {
+    /// An empty memo field.
+    Empty,
+    /// A memo field containing a UTF-8 string.
+    Text(TextMemo),
+    /// Some unknown memo format from ✨*the future*✨ that we can't parse.
+    Future(MemoBytes),
+    /// A memo field containing arbitrary bytes.
+    Arbitrary(Box<[u8; 511]>),
+}
+
+impl fmt::Debug for Memo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Memo::Empty => write!(f, "Memo::Empty"),
+            Memo::Text(memo) => write!(f, "Memo::Text(\"{}\")", memo.0),
+            Memo::Future(bytes) => write!(f, "Memo::Future({:0x})", bytes.0[0]),
+            Memo::Arbitrary(bytes) => {
+                write!(f, "Memo::Arbitrary(")?;
+                fmt_colon_delimited_hex(f, &bytes[..])?;
+                write!(f, ")")
+            }
+        }
+    }
+}
+
+impl Default for Memo {
+    fn default() -> Self {
+        Memo::Empty
+    }
+}
+
+impl PartialEq for Memo {
+    fn eq(&self, rhs: &Memo) -> bool {
+        match (self, rhs) {
+            (Memo::Empty, Memo::Empty) => true,
+            (Memo::Text(a), Memo::Text(b)) => a == b,
+            (Memo::Future(a), Memo::Future(b)) => a.0[..] == b.0[..],
+            (Memo::Arbitrary(a), Memo::Arbitrary(b)) => a[..] == b[..],
+            _ => false,
+        }
+    }
+}
+
+impl TryFrom<MemoBytes> for Memo {
+    type Error = Error;
+
+    /// Parses a `Memo` from its ZIP 302 serialization.
+    ///
+    /// Returns an error if the provided slice does not represent a valid `Memo` (for
+    /// example, if the slice is not 512 bytes, or the encoded `Memo` is non-canonical).
+    fn try_from(bytes: MemoBytes) -> Result<Self, Self::Error> {
+        match bytes.0[0] {
+            0xF6 if bytes.0.iter().skip(1).all(|&b| b == 0) => Ok(Memo::Empty),
+            0xFF => Ok(Memo::Arbitrary(Box::new(bytes.0[1..].try_into().unwrap()))),
+            b if b <= 0xF4 => str::from_utf8(bytes.as_slice())
```

Comment:
I just noticed that this doesn't yet implement the `0xF5` case specified in ZIP 302. That specification also needs some clarification related to how to parse the `length` bytes (see https://github.com/zcash/zips/pull/105#discussion_r596966361).

Reply:
I intentionally removed 0xF5 handling from this PR last year, in an effort to make the PR mergeable, as it was the most contentious part.